src/unicode.cpp

```diff
@@ -9,6 +9,39 @@
 
 namespace ada::unicode {
 
+  /**
+   * The has_tabs_or_newline function is a bottleneck and it is simple enough that compilers
+   * like GCC can 'autovectorize it'.
+   */
+  ada_really_inline constexpr bool has_tabs_or_newline(std::string_view user_input) noexcept {
+    auto has_zero_byte = [](uint64_t v) {
+      return ((v - 0x0101010101010101) & ~(v)&0x8080808080808080);
+    };
+    auto broadcast = [](uint8_t v) -> uint64_t { return 0x101010101010101 * v; };
+    size_t i = 0;
+    uint64_t mask1 = broadcast('\r');
+    uint64_t mask2 = broadcast('\n');
+    uint64_t mask3 = broadcast('\t');
+    uint64_t running{0};
+    for (; i + 7 < user_input.size(); i += 8) {
```

Comment:
Shouldn’t there be an exit condition in here, to avoid iterating through the whole input?

Reply:
It depends on what your expectation is. Do you expect that the code will run through it all most of the time? Then that's what you want to optimize for. The compiler is able to vectorize the code because it is super simple (GCC 12 will produce SSE SIMD code when compiler with -O3). If you add a branch, it is not longer easily vectorizable and the compiler won't even try.

Even when you work it out by hand, branching will come at a slight penalty. Here, we don't have the luxury of working it out by hand since we rely on the compiler.